### PR TITLE
Global state and authentication

### PR DESCRIPTION
**Pull request by Vamania69**
set up the global user cart state and add to cart functionality at local state and also at the database

### CODE REVIEW

File: src/components/RegisterUser.tsx [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    setRegisterUserDetails({
+      ...RegisterUserDetails,
+      [e.target.name]: e.target.value,
+    });
+    console.log(RegisterUserDetails);
+  };
+
+  const SubmitHandler = async () => {
+    try {
+      const response = await axios.post(
```
Comment (AnadiSharma007) [inline src/components/RegisterUser.tsx]: you can create get, post, put, delete wrappers in the axios.instance

File: src/components/RegisterUser.tsx [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+  Heading,
+  Input,
+  Stack,
+  Image,
+} from "@chakra-ui/react";
+import axios from "axios";
+import { BASE_URL } from "../utils/constants";
+import Link from "next/link";
+export default function RegisterUser() {
+  const [RegisterUserDetails, setRegisterUserDetails] = useState({
```
Comment (AnadiSharma007) [inline src/components/RegisterUser.tsx]: variable name should start with small letter and must follow camalCasing.

File: src/components/SignIn.tsx [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+    <Stack minH={"100vh"} direction={{ base: "column", md: "row" }}>
+      <Flex p={8} flex={1} align={"center"} justify={"center"}>
+        <Stack spacing={4} w={"full"} maxW={"md"}>
+          <Heading fontSize={"2xl"}>Sign in to your account</Heading>
+          <FormControl id="email">
+            <FormLabel>Email address</FormLabel>
+            <Input
+              name="userEmail"
+              type="email"
+              value={loginDetails.userEmail}
```
Comment (AnadiSharma007) [inline src/components/SignIn.tsx]: added ?.

File: src/globalReduxStore/features/cart/cartStoreSlice.ts [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+// this is the initial state of the cart which is empty array 
+let initialState: userState = [];
+
+
+// update function that will update the cart in the database
+
+ const updateCart = async (productId : string)=>{
+   // get the product id and send it to the server as a object 
+   const response = await post("/user/cart",{productId : productId })
```
Comment (AnadiSharma007) [inline src/globalReduxStore/features/cart/cartStoreSlice.ts]: if your key and value are same such as {productId: productId}. Then you can represent it as {productId}.

File: src/globalReduxStore/features/cart/cartStoreSlice.ts [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+      // check if the cart is empty && set the cart initial state
+      if (state.length === 0) {
+        state.push(action.payload);
+      }
+      // check weather product exits in the cart or not
+      else {
+        if (
+          state.find(
+            (product) =>
+              product.product.productId === action.payload.product.productId
```
Comment (AnadiSharma007) [inline src/globalReduxStore/features/cart/cartStoreSlice.ts]: add ?.

File: src/pages/products/[productId].tsx [hunk 6]
```diff
@@ -33,7 +34,10 @@ const ProductDescription = () => {
   }, [isRouterReady]);
 
   return isLoading ? (
-    <h1 className="text-center">Loading...</h1>
+    <div className="text-center">
```
Comment (AnadiSharma007) [inline src/pages/products/[productId].tsx]: You can create a shared component for loding state.

File: src/pages/cart.tsx [hunk 7]
```diff
@@ -1,42 +1,63 @@
...
+  const [userCart, setUserCart] = useState(cartProducts);
 
-import React from 'react'
+  useEffect(() => {
+    // updating the local state of the user login
+    setIsUserLoggedIn(userLoggedInStatus);
+  }, [userLoggedInStatus, cartProducts]);
+  
+  
+  useEffect(() => {
```
Comment (AnadiSharma007) [inline src/pages/cart.tsx]: do you need this empty useEffect?

File: src/components/RegisterUser.tsx (hunk 1 above)
Comment (Vamania69) [reply src/components/RegisterUser.tsx]: already made but but forget to use

File: src/globalReduxStore/features/cart/cartStoreSlice.ts (hunk 5 above)
Comment (Vamania69) [reply src/globalReduxStore/features/cart/cartStoreSlice.ts]: is this look like this  product.product?.productId=== action?.payload?.product?.productId  and like wise every where will be checked

File: src/pages/products/[productId].tsx (hunk 6 above)
Comment (Vamania69) [reply src/pages/products/[productId].tsx]: created shared isLoadingComponent

**Comment by Vamania69** [comment]
sir before this pr could be merged we need to resolve the issues and mentioned changes which i could not resolve in #6 . then we will merge the pr #6 then this could be merged as per my understanding .

**Pull request merged**